### DISCUSSION TITLE
How to properly use PartitionedConfig with Python config?

### DISCUSSION THREAD

**Discussion by rprechelt** [dagster-io/dagster#18626] [Q&A] [answered]
Is there a complete example for how to use `PartitionedConfig` correctly? As far as I'm aware, there isn't one in the docs or the example projects. I've tried to use it as demonstrated below but when attempting to execute a job through the Dagster UI, I'm getting the following exception so I suspect I'm doing something wrong. The header of the pop-up box is _"Unable to load tags"._ 

If I change the `config` argument to be a hard-coded `RunConfig({"my_asset": MyConfig()})` then the job will run successfully. 

```
TypeError: Object of type MyConfig is not JSON serializable
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/site-packages/dagster/_grpc/server.py", line 534, in ExternalPartitionConfig
    serialized_data = serialize_value(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/site-packages/dagster/_serdes/serdes.py", line 617, in serialize_value
    return seven.json.dumps(packed_value, **json_kwargs)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/opt/homebrew/Caskroom/miniconda/base/envs/py310_64/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
```

Here is the relevant snippets:

```python
partitions = StaticPartitionsDefinition(["red", "blue", "green"])

class MyConfig(Config):
    x: float = 1

@asset(partitions_def=partitions)
def my_asset(config: MyConfig):
    return config.x

MyConfig().model_dump_json() # this works and converts to JSON

def run_config(key: str) -> Mapping[str, Any]:
    return {"my_asset": MyConfig()}

config = PartitionedConfig(partitions=partitions, run_config_for_partition_key_fn=run_config)

config.get_run_config_for_partition_key("red") # works and return {"my_asset": MyConfig(x=1)}.

# running this job gives the "Unable to load tags" error
job1 = define_asset_job(name="my_job1", partitions_def=partitions, config=config, selection=["my_asset"])

# this job runs fine
job2 = define_asset_job(name="my_job2", partitions_def=partitions, config=RunConfig({"my_asset": MyConfig()}), selection=["my_asset"])

```

**Comment by jamiedemaria** [comment]
`PartitionedConfig` is mainly used to make it possible to model partitions using ops and op-based jobs. The function decorators `static_paritioned_config`, `date_partitioned_config` etc, are helper decorators that created `PartitionedConfig` for you. You can see examples of their usage [here](https://docs.dagster.io/concepts/partitions-schedules-sensors/partitioning-ops#defining-jobs-with-static-partitions). But like I mentioned, these decorators are there that ops can be written to represent partitions. Since assets support partitions more natively, `PartitionedConfig` is not required. What are you trying to achieve by using `PartitionedConfig`? There might be a different object you need to use

**Reply by rprechelt** [reply]
Thanks for the quick response! 

I wish to set the config of an asset differently _per_ partition _key_, i.e.

```
partitions = StaticPartitionsDefinition(["red", "blue", "green"])

@asset(partitions_def=partitions)
def my_asset(config: MyConfig):
    pass
```

I want the `config` _instance_ passed to the asset func to be unique to each partition key, e.g. a different `config` for `red`, `blue`, and `green`. From reading the docs, this seemed like exactly what `PartitionedConfig` was intended for: provide a function, `run_config` in my example above, that can be used to create the config to be used for a particular partition key, but I do understand now that it was more intended for use with ops.

If there is another way to achieve that, fantastic. The docstring for `define_asset_job` does explicitly list `PartitionedConfig` as an allowed value for the `config` argument so that appeared to be the recommended solution (and I don't understand the origin of the "Unable to load tags"/"JSON" error that I originally posted).

FYI: In reality, my asset has >100M unique partition keys (not 3 like in my example) so any "hand-coded" solution for each asset key is unworkable—it really needs to be looked up via a function.

**Reply by jamiedemaria** [reply, marked as answer]
I dug into this a bit more, and yeah you're right that `PartitionedConfig` should be usable. I think the nuance is that the return value from `run_config_for_partition_key_fn` needs to be a fully resolved config dictionary, not a subclass of `Config`.

Some history: Originally, configuration was only ever specified as dictionaries. In a more recent version of Dagster, we introduced the Pythonic config system, which allows for the use of pydantic-style configuration classes. These classes are then converted to dictionaries, and that's why the serialization is failing - because an instance of `MyConfig` isn't serializable by the seven library. 

I believe if you modified the function to return the dictionary, it should work:
```python
def run_config(key: str) -> Mapping[str, Any]:
   return {
        "ops": {
            "my_asset": {"config": {"x": 2}}
        }
    }
```

we have an issue tracking allowing `PartitionedConfig` to support pydantic config here https://github.com/dagster-io/dagster/issues/16812

**Reply by rprechelt** [reply]
Thanks @jamiedemaria! Changing my function to return the fully resolved config dictionary gets everything working.

Do you accept PRs for documentation changes from non-Dagster contributors? An example of the proper return type of `run_config_for_partition_key_fn` in the docs for `PartitionedConfig` docs would have saved me hours of troubleshooting.

**Reply by jamiedemaria** [reply]
yup - feel free to submit a PR for docs changes!
